[Inductor] Update DeviceAssert op to behave like store

karthickai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163696

Updated the DeviceAssert operation to match the behavior of Store, it will fixes the issue mentioned in [this PR](https://github.com/pytorch/pytorch/pull/163023) and updated testcases as Elias [suggested](https://github.com/pytorch/pytorch/pull/160677#discussion_r2353834646).



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos @choijon5

karthickai committed:
[Inductor] Update DeviceAssert op to behave like store

[ghstack-poisoned]

mlazos commented on torch/_inductor/codegen/common.py:
```diff
@@ -2118,6 +2123,9 @@ def store(
         raise NotImplementedError
 
+    def device_assert_async(self, cond: CSEVariable, msg: str) -> None:
+        raise NotImplementedError
```
nit: make this have a message like the other one?

mlazos commented on torch/_inductor/ops_handler.py:
```diff
@@ -1158,3 +1149,6 @@ def _default(self, name: str, args: tuple[Any, ...], kwargs: dict[str, Any]) ->
         return val
+
+    def device_assert_async(self, *args, **kwargs) -> None:
+        raise NotImplementedError("device_assert_async not implemented")
```
similar comment, message consistency?

karthickai committed:
Update on "[Inductor] Update DeviceAssert op to behave like store"


Updated the DeviceAssert operation to match the behavior of Store, it will fixes the issue mentioned in [this PR](https://github.com/pytorch/pytorch/pull/163023) and updated testcases as Elias [suggested](https://github.com/pytorch/pytorch/pull/160677#discussion_r2353834646).



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos choijon5 

[ghstack-poisoned]

karthickai commented:
@pytorchbot merge
